### DISCUSSION TITLE
Plugin for vi-like modes

### DISCUSSION THREAD

**Discussion by lryffel** [xournalpp/xournalpp#2931] [Q&A] [answered]
I would very much like my xournal++ configuration to feel like vim. Luckily, by using the Lua plugins I can map, e..g, `w` to the pen tool, `e` to the eraser tool, etc. In this way, I can leave my left hand firmly on the left half of the keyboard and can switch between the tools lightning fast.

What I cannot achieve so far are the famous "modes" of vim, in particular insert mode. In order to enter text into a text field, I would like to be able to press `i` and have this disable all my accelerators. I can only achieve this partially (and uselessly) by only executing the command in normal mode. Here's (a minimal version of) what I got so far:

```lua
function initUi()
    -- mode shortcuts
  app.registerUi({["menu"] = "Enter insert mode", ["callback"] = "insert_mode", ["accelerator"] = "i"}); 
  app.registerUi({["menu"] = "Enter normal mode", ["callback"] = "normal_mode", ["accelerator"] = "<Ctrl><Shift>c"}); 
  -- normal mode
  app.registerUi({["menu"] = "Select pen tool (normal mode)", ["callback"] = "pen", ["accelerator"] = "w"}); 
  app.registerUi({["menu"] = "Select eraser tool (normal mode)", ["callback"] = "eraser", ["accelerator"] = "e"}); 
end



-- MODES
-- the modes are "normal", "insert", and "color"
local currentMode = "normal"

function insert_mode()
  if currentMode == "normal" then
    currentMode = "insert"
  else
    -- I want to send "i" to xournal here
  end
end

function normal_mode()
  currentMode = "normal"
end



-- TOOLS
function pen()
  if currentMode == "normal" then
    app.uiAction({["action"] = "ACTION_TOOL_PEN"})
  else
    -- I want to send "w" to xournal here
  end
end

function eraser()
  if currentMode == "normal" then
    app.uiAction({["action"] = "ACTION_TOOL_ERASER"})
  else
    -- I want to send "e" to xournal here
  end
end
```

The comments should explain what I want to do: I want to send a keypress signal to xournal++ without it getting intercepted by my Lua script. Is this possible at all?

In case not, I should add that the modes would not only be useful for inserting text into textboxes, but also for selecting colours (e.g., selecting red by pressing `cr`, selecting yellow by pressing `cy`, etc.) Are there alternative solutions that do not require forwarding of keypresses but still allow me to have this mode-like behaviour?

**Comment by rolandlo** [comment, marked as answer]
I see what you want to do. The intercepted keypresses is a problem that we have to solve in Xournal++ itself. I don't think it can be done in the plugin (at least with the current plugin API). See #2755 and #977 for some discussions around this problem.

**Reply by lryffel** [reply]
I guess it is sort of what I expected. Although [this](https://github.com/xournalpp/xournalpp/issues/977#issuecomment-731657423) seems promising, I don't quite know what to do with it. Since TeX mode does not seem to be affected by the Lua accelerators, this is  actually a reasonable substitute for my insert mode. Thanks again for your help!

**Comment by lryffel** [comment]
[Here](https://github.com/raw-bacon/vi-xournalpp)'s what I ended up making. Can't forward key presses, so it breaks some features, but otherwise seems to work pretty well.

**Reply by rolandlo** [reply]
Looks nice and clean!
